### ISSUE TITLE
v.1.1.352 breaks on node.js v12

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#7365]
**Describe the bug**

The latest release [fails to run on node.js v12](https://github.com/RobertCraigie/pyright-python/actions/runs/8088861823/job/22103745277?pr=252) due to a module not found error:
```
internal/modules/cjs/loader.js:818
  throw err;
  ^

Error: Cannot find module 'node:util'
Require stack:
- /home/runner/.cache/pyright-python/1.1.352/node_modules/pyright/dist/pyright.js
- /home/runner/.cache/pyright-python/1.1.352/node_modules/pyright/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:815:15)
    at Function.Module._load (internal/modules/cjs/loader.js:667:27)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.9632 (/home/runner/.cache/pyright-python/1.1.352/node_modules/pyright/dist/pyright.js:1:610)
    at o (/home/runner/.cache/pyright-python/1.1.352/node_modules/pyright/dist/pyright.js:1:1243)
    at Object.1264 (/home/runner/.cache/pyright-python/1.1.352/node_modules/pyright/dist/vendor.js:2:794958)
    at o (/home/runner/.cache/pyright-python/1.1.352/node_modules/pyright/dist/pyright.js:1:1243)
    at Object.1476 (/home/runner/.cache/pyright-python/1.1.352/node_modules/pyright/dist/pyright-internal.js:1:441151)
    at o (/home/runner/.cache/pyright-python/1.1.352/node_modules/pyright/dist/pyright.js:1:1243) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/runner/.cache/pyright-python/1.1.352/node_modules/pyright/dist/pyright.js',
    '/home/runner/.cache/pyright-python/1.1.352/node_modules/pyright/index.js'
  ]
}
```

Node.js v16 tests pass, I haven't tested v14.

I don't actually know if any pyright-python users are relying on node.js v12 so if you want to drop support for v12 that'd probably be a fine solution, I can update pyright-python accordingly.

**VS Code extension or command-line**

`v.1.1.352`

**Comment by erictraut** [comment]
I don't think I've made any changes that would cause this, so I suspect this incompatibility came in from the pylance team. @debonte or @rchiodo, could you please investigate and advise?

**Comment by debonte** [comment]
We upgraded to node.js v18 recently -- https://github.com/microsoft/pyright/commit/e955fab18e5a2994c83d5909bbb644705dd4eb75

**Comment by erictraut** [comment]
Rich assured me that the v18 upgrade was only for the VS Code extension and the build environment — that it would not affect the CLI compatibility with v12. If it does, then that was a miss, and we should discuss.

**Comment by rchiodo** [comment]
I can only guess that building with node 18 vs node 16 is causing this problem. The engine target is still 12.

**Comment by erictraut** [comment]
I built it with v16 this time, so that doesn't explain it.

**Comment by rchiodo** [comment]
Because pyright doesn't seem to reference that itself.

**Comment by rchiodo** [comment]
With debug bits I can see the culprit. It's the new vscode-languageserver bits:

```
Error: Cannot find module 'node:util'
Require stack:
- C:\Users\rchiodo\source\repos\pyright\packages\pyright\dist\pyright.js
- C:\Users\rchiodo\source\repos\pyright\packages\pyright\index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:815:15)
    at Function.Module._load (internal/modules/cjs/loader.js:667:27)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.node:util (C:\Users\rchiodo\source\repos\pyright\packages\pyright\dist\pyright.js:121:18)
    at __webpack_require__ (C:\Users\rchiodo\source\repos\pyright\packages\pyright\dist\pyright.js:255:42)
    at Object.../pyright-internal/node_modules/vscode-languageserver/lib/node/main.js (C:\Users\rchiodo\source\repos\pyright\packages\pyright\dist\vendor.js:34509:21)
    at __webpack_require__ (C:\Users\rchiodo\source\repos\pyright\packages\pyright\dist\pyright.js:255:42)
    at Object.../pyright-internal/src/analyzer/program.ts (C:\Users\rchiodo\source\repos\pyright\packages\pyright\dist\pyright-internal.js:26184:33)
    at __webpack_require__ (C:\Users\rchiodo\source\repos\pyright\packages\pyright\dist\pyright.js:255:42) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'C:\\Users\\rchiodo\\source\\repos\\pyright\\packages\\pyright\\dist\\pyright.js',
    'C:\\Users\\rchiodo\\source\\repos\\pyright\\packages\\pyright\\index.js'
  ]
}
```

**Comment by rchiodo** [comment]
When we moved to `18` we also upgraded the vscode-languageserver bits. This commit here brought in `node:util`:

https://github.com/microsoft/vscode-languageserver-node/commit/a57d7f12ae37e66f6c56297eab1d12a3a4ccf92c

**Comment by rchiodo** [comment]
Maybe we can fake it to use something else, but I think we'll have to upgrade to 14 if we want to support the latest vscode-languageserver protocols. We needed this to support non COI cancellation in the browser for Pylance. AFAIK, Pylance and Pyright have to be in sync with respect to their vscode-languageserver versions.

**Comment by debonte** [comment]
> if we want to support the latest vscode-languageserver protocols...

FYI, the latest vscode-languageserver-node release (from [today](https://github.com/microsoft/vscode-languageserver-node/tags)) has a [bug fix](https://github.com/microsoft/vscode-languageserver-node/pull/1390) that I've been waiting for. I was planning to upgrade after the upcoming Pylance stable release ships.

**Comment by erictraut** [comment]
I think it's fine to update the minimum node version from v12 to v14. After all, v12 is really ancient at this point.

**Comment by RobertCraigie** [comment]
Dropping support for v12 is fine with me, thanks for investigating!

**Comment by erictraut** [comment]
This is included in pyright 1.1.353, which I just published.

**Comment by acampove** [comment]
Except that the newest version of nodejs that ubuntu users can install through apt is `v12.22.9`, at least after doing:

```
  662  sudo apt update
  663  sudo apt install nodejs
  664  node -v
  665  sudo apt upgrade nodejs
  666  node -v
```

I am using `Ubuntu 22.04.3 LTS`. People keep telling me to move to nix, I wonder why the default package manager prefers to install an ancient version of this nodejs.
